Android asynchronous rendering

ivovandongen opened the pull request:
Builds on #8820 to implement asynchronous rendering on Android

Pre-requisites:
- [x] Make `thread.hpp` public
- [x] Make `Thread::pause/resume` thread safe. So we can pause resume from arbitrary threads (main/opengl).
- [x] Allow safe direct access to Actor's managed object through `Thread`. We need to be able to pause the scheduler and invoke the object (`Renderer`) from another thread (OpenGl renderer thread).
- [x] Add ask pattern to `Actor`/`ActorRef` to make blocking calls from the main thread to the orchestration thread (`query*Features`) https://github.com/mapbox/mapbox-gl-native/pull/9591
- [x] Make self reference on Actors optional so we can make the `Renderer` an `Actor` without modifying it. https://github.com/mapbox/mapbox-gl-native/pull/9591

Steps:
- [x] Remove texture view support for now
- [x] Separate `RenderBackend` and `View` from `NativeMapView` and make it owned by the `RendererFrontend` to make locking and dispatching more contained.
- [x] Introduce orchestration thread + synchronization
- [x] Add `GLSurfaceview`
- [x] Render from OpenGl thread
- [x] Handle GLSurfaceView life-cycle
- [x] Move EGL / OpenGL management code to Java
- [x] Handle snapshots
- [x] Handle fps logging
- [ ] (optionally) Implement Asynchronous rendering with `TextureView`


Fixes #9956

ivovandongen commented:
@jfirebaugh Making some headway, but ran into an interesting crash:

<details>
  <summary>Crash dump</summary>
<pre>
********* Crash dump: **********
Build fingerprint: 'google/sailfish/sailfish:7.1.2/NJH47D/4045516:user/release-keys'
pid: 21745, tid: 22301, name: GLThread 5777  >>> com.mapbox.mapboxsdk.testapp <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x24
Stack frame #00 pc 0025d97a  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine std::__ndk1::unique_ptr<mbgl::util::RunLoop::Impl, std::__ndk1::default_delete<mbgl::util::RunLoop::Impl> >::get() const at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2725
Stack frame #01 pc 002675db  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine Impl at /Users/ivo/git/mapbox-gl-native/platform/android/src/timer.cpp:57
Stack frame #02 pc 002671df  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine std::__ndk1::__unique_if<mbgl::util::Timer::Impl>::__unique_single std::__ndk1::make_unique<mbgl::util::Timer::Impl>() at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:3149
Stack frame #03 pc 003f9bb7  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine Throttler at /Users/ivo/git/mapbox-gl-native/src/mbgl/util/throttler.cpp:6
Stack frame #04 pc 0037f797  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine GeometryTile at /Users/ivo/git/mapbox-gl-native/src/mbgl/tile/geometry_tile.cpp:45
Stack frame #05 pc 002e28e5  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine AnnotationTile at /Users/ivo/git/mapbox-gl-native/src/mbgl/annotation/annotation_tile.cpp:13
Stack frame #06 pc 005974ef  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine std::__ndk1::__unique_if<mbgl::AnnotationTile>::__unique_single std::__ndk1::make_unique<mbgl::AnnotationTile, mbgl::OverscaledTileID const&, mbgl::TileParameters const&>(mbgl::OverscaledTileID const&, mbgl::TileParameters const&) at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:3149
Stack frame #07 pc 00597355  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine _ZNSt6__ndk18__invokeIRZN4mbgl22RenderAnnotationSource6updateENS1_9ImmutableINS1_5style6Source4ImplEEERKNS_6vectorINS3_INS4_5Layer4ImplEEENS_9allocatorISB_EEEEbbRKNS1_14TileParametersEE3$_0JRKNS1_16OverscaledTileIDEEEEDTclclsr3std6__ndk1E7forwardIT_Efp_Espclsr3std6__ndk1E7forwardIT0_Efp0_EEEOSP_DpOSQ_ at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/type_traits:4287
Stack frame #08 pc 005972a1  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine std::__ndk1::__function::__func<mbgl::RenderAnnotationSource::update(mbgl::Immutable<mbgl::style::Source::Impl>, std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&)::$_0, std::__ndk1::allocator<mbgl::RenderAnnotationSource::update(mbgl::Immutable<mbgl::style::Source::Impl>, std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&)::$_0>, std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>::operator()(mbgl::OverscaledTileID const&) at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/functional:1533
Stack frame #09 pc 0059c321  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>::operator()(mbgl::OverscaledTileID const&) const at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/functional:1896
Stack frame #10 pc 0059bd4d  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine operator() at /Users/ivo/git/mapbox-gl-native/src/mbgl/renderer/tile_pyramid.cpp:140
Stack frame #11 pc 00599a9d  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine void mbgl::algorithm::updateRenderables<mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_0, mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_1, mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_2, mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_4, std::__ndk1::vector<mbgl::UnwrappedTileID, std::__ndk1::allocator<mbgl::UnwrappedTileID> > >(mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_0, mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_1, mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_2, mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>)::$_4, std::__ndk1::vector<mbgl::UnwrappedTileID, std::__ndk1::allocator<mbgl::UnwrappedTileID> > const&, mbgl::Range<unsigned char> const&, unsigned char) at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../src/mbgl/algorithm/update_renderables.hpp:37
Stack frame #12 pc 00598491  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::TilePyramid::update(std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&, mbgl::SourceType, unsigned short, mbgl::Range<unsigned char>, std::__ndk1::function<std::__ndk1::unique_ptr<mbgl::Tile, std::__ndk1::default_delete<mbgl::Tile> > (mbgl::OverscaledTileID const&)>) at /Users/ivo/git/mapbox-gl-native/src/mbgl/renderer/tile_pyramid.cpp:162
Stack frame #13 pc 00596c49  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::RenderAnnotationSource::update(mbgl::Immutable<mbgl::style::Source::Impl>, std::__ndk1::vector<mbgl::Immutable<mbgl::style::Layer::Impl>, std::__ndk1::allocator<mbgl::Immutable<mbgl::style::Layer::Impl> > > const&, bool, bool, mbgl::TileParameters const&) at /Users/ivo/git/mapbox-gl-native/src/mbgl/annotation/render_annotation_source.cpp:35
Stack frame #14 pc 00482afd  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::RenderStyle::update(mbgl::UpdateParameters const&) at /Users/ivo/git/mapbox-gl-native/src/mbgl/renderer/render_style.cpp:226
Stack frame #15 pc 002f507b  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::Renderer::Impl::render(mbgl::View&, mbgl::UpdateParameters const&) at /Users/ivo/git/mapbox-gl-native/src/mbgl/renderer/renderer_impl.cpp:53
Stack frame #16 pc 002f21cd  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::Renderer::render(mbgl::View&, mbgl::UpdateParameters const&) at /Users/ivo/git/mapbox-gl-native/src/mbgl/renderer/renderer.cpp:25
Stack frame #17 pc 002ac0c1  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::android::AndroidRendererFrontend::render() at /Users/ivo/git/mapbox-gl-native/platform/android/src/android_renderer_frontend.cpp:119
Stack frame #18 pc 000b991d  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::android::NativeMapView::render(_JNIEnv&) at /Users/ivo/git/mapbox-gl-native/platform/android/src/native_map_view.cpp:189
Stack frame #19 pc 000cd4cb  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine operator() at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:245
Stack frame #20 pc 000cd461  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine operator() at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:112
Stack frame #21 pc 000cd429  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine __invoke at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:110
Stack frame #22 pc 000cd555  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine operator()<jni::jobject *> at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:59
Stack frame #23 pc 000cd515  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine __invoke<jni::jobject *> at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:55
Stack frame #24 pc 0054ede7  /data/app/com.mapbox.mapboxsdk.testapp-2/oat/arm/base.odex (offset 0x524000)
</pre>
</details>



It seems like updating the render style from the render thread is not working as the placement throttler is trying to access the `Runloop` which doesn't exist on the Render thread. 

We could install a runloop on the render thread and call `runOnce()` on every render iteration, but I doubt that is going to work as we want. Alternative is to split `Renderer::update` of from `Renderer::render` and make sure the former is only executed on the orchestration thread and the latter on the render thread.

jfirebaugh commented:
Could we move the throttling into the worker itself? Or is the important thing to throttle actually the worker invocation itself? cc @ChrisLoer

ivovandongen commented:
> Could we move the throttling into the worker itself

If throttling is the only blocker here, we could probably find a solution. I'm wondering a bit if there are no catches when updating from the render thread. But I'll hack it up a bit for now to see if anything else comes up.

ChrisLoer commented:
Sending a bunch of unused invocations to the worker mailbox shouldn't be a big deal if the handling is throttled by the worker. But you still have to somehow self-send a "do placement once my throttling duration is finished" message, which needs something like `Timer`/`RunLoop`.

jfirebaugh commented:
> If throttling is the only blocker here, we could probably find a solution. I'm wondering a bit if there are no catches when updating from the render thread.

You're right, it's not going to be the only blocker -- `GeometryTile` also assumes there is a `RunLoop` when creating the mailbox for messages coming back from the worker.

Will think about this more.

jfirebaugh commented:
I think the cleanest way to resolve this is by avoiding the need for an orchestration/renderer thread distinction and have only the renderer thread. To do this we need to fix code that currently assumes that it's running on a thread with a `RunLoop`:

* Replace the current implicitly global "current `RunLoop`" with an implicitly global "current `Scheduler`": expose a static `Scheduler::current()` and use that for constructing `Mailbox`es instead of `RunLoop:::Get()`.
* `AndroidRendererFrontend` should replace `util::Thread<Renderer>` with a custom `Scheduler` implementation that schedules onto the `GLSurfaceView` thread using [`GLSurfaceView::queueEvent`](https://developer.android.com/reference/android/opengl/GLSurfaceView.html#queueEvent(java.lang.Runnable)). For the renderer thread, this will be the scheduler returned by `Scheduler::current()`.
* That still leaves `Throttler`/`Timer`. Likely what needs to happen there is that both are rewritten in terms of actors instead of in terms of `std::function` and `RunLoop`. We should look into how this is handled [in akka](http://doc.akka.io/docs/akka/current/java/scheduler.html) or other actor systems.

ivovandongen commented:
@jfirebaugh I've been thinking a bit more about using only the render thread instead of a three thread setup. I like the simplification, but see one possible issue with this; it would probably make any background work impossible. The processing of the `RenderStyle` would be tightly coupled to the lifecycle of the visible parts of the map as the GL Thread is stopped when the containing activity is paused (not the current activity / the app is backgrounded). This would mainly be a problem for offline downloads, but would also limit continuation of resource loading while the  user switches to other apps.

From the [documentation](https://developer.android.com/reference/android/opengl/GLSurfaceView.html):

> A GLSurfaceView must be notified when to pause and resume rendering. GLSurfaceView clients are required to call onPause() when the activity stops and onResume() when the activity starts. These calls allow GLSurfaceView to pause and resume the rendering thread, and also allow GLSurfaceView to release and recreate the OpenGL display.

This is in line with what @kkaefer previously wrote up about the GLSurface view: https://github.com/mapbox/mapbox-gl-native/issues/5766#issuecomment-300180689

Another complication is continued support of [TextureView](https://developer.android.com/reference/android/view/TextureView.html). Needed for certain use cases where mapviews are layered on top of other elements. Although we need to do the threading here ourselves, it would be great not to diverge too much for maintainability.

I'll verify first what the actual behaviour is of the GL Thread when the activity is paused and if any work can be scheduled before continuing here.

ivovandongen commented:
> I'll verify first what the actual behaviour is of the GL Thread when the activity is paused and if any work can be scheduled before continuing here.

I've setup a basic Activity that continuously schedules on the GL Thread using [GLSurfaceView#queueEvent](https://developer.android.com/reference/android/opengl/GLSurfaceView.html#queueEvent(java.lang.Runnable)). When paused, the work continues to execute on the test phones I have. Not sure if we can rely on that behaviour on all phones we support though.

jfirebaugh commented:
Yes, good point. Even if we do keep two threads though, I think the actions to get that working are largely the same: removing reliance on a per-thread `RunLoop`.

jfirebaugh commented:
> The processing of the `RenderStyle` would be tightly coupled to the lifecycle of the visible parts of the map as the GL Thread is stopped when the containing activity is paused (not the current activity / the app is backgrounded).

This actually seems like the desired behavior to me. When the activity is paused, rendering work should also pause, no?

> This would mainly be a problem for offline downloads, but would also limit continuation of resource loading while the  user switches to other apps.

Offline downloads happen on a different thread though, so they shouldn't be affected.

ivovandongen commented:
@jfirebaugh In https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/a3afcb546b165b96ecc631dbbd6e88bfcf53b103 I've implemented the async rendering with 2 threads (dropping the Orchestration thread).

Since the GLThread has no runloop, and cannot have a runloop as it blocks on a mutex when used in `RENDERMODE_WHEN_DIRTY` until `GLSurfaceView#requestRender` is called, I've added a scheduler implementation that calls back into the main thread to request a render whenever `Scheduler::schedule` is called and processes the mailbox when a render is actually performed.

This doesn't, however, work when we need to make a blocking call into the renderer (query*Features for example). On the main thread the request is made through `ActorRef::ask` and then blocks on a future to await the result. The GLThread requests a render to process the request, but since the main thread is blocked, this will never happen and the main thread remains block while the GLThread is halted at it's mutex. 

https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/82f03a9bb57a0a523291e5340206cb5b873a8415 addresses that by skipping the actor for callbacks and the async task in between. This does mean that the invalidation is now done by two threads and the calls from the gl thread are not coalesced anymore. Need to investigate if this hampers performance.

https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/117b8997778e5d4beb14e45cc69b3b3ac2356721 adds a mutex to guard the update parameters. This solves a lot of the crashes I've been seeing (should have seen this before), but also fixes the flickering we saw on pinch gestures.

There are still a couple of crashes left, mostly around annotations (bitmaps) it seems. I'm going to dig into that now.

Do you still prefer this direction, with two threads over the 3 thread approach? It seems to work well for the most part now, I'm not sure if it is much cleaner as it is less obvious which parts are running on which thread(s).

jfirebaugh commented:
Did you try the approach we discussed of using `GLSurfaceView.queueEvent` to implement `schedule?

ivovandongen commented:
> Did you try the approach we discussed of using GLSurfaceView.queueEvent to implement schedule?

https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/4a6d067e7cf19c1445770f5fbee08750505a79ed Uses `GLSurfaceView#queueEvent` to schedule work items and process them separately. 

Although it might be a nice abstraction (esp when we re-introduce a form of the `TextureView`), it doesn't fundamentally change anything. Only difference is that before the whole queue on the scheduler is processed on a render call before starting the actual render and with this commit the work is scheduled separately from rendering. This might introduce a slight overhead in JNI calls as there are more calls to do the same amount of work, but this is probably negligible.

ivovandongen commented:
> There are still a couple of crashes left, mostly around annotations (bitmaps) it seems. 

Separate issue: https://github.com/mapbox/mapbox-gl-native/issues/9936

ivovandongen commented:
@jfirebaugh I've been testing more extensively and I'm still running into deadlocks with both 2-thread solutions. Mostly notably, when destructing the `Map` while resources are being loaded (or any other background activity is happening) I'm running into situations like these:

<details>
 <summary>Main thread blocked on `Mailbox::close()`</summary>

```
"main" prio=5 tid=1 Native
| group="main" sCount=1 dsCount=0 flags=1 obj=0x73c9f568 self=0xe5404000
| sysTid=12391 nice=-10 cgrp=default sched=0/0 handle=0xe8e814a8
| state=S schedstat=( 2624847218 273684725 2360 ) utm=247 stm=14 core=2 HZ=100
| stack=0xfeffa000-0xfeffc000 stackSize=8MB
| held mutexes=
kernel: __switch_to+0xb0/0xbc
kernel: futex_wait_queue_me+0xdc/0x168
kernel: futex_wait+0xf4/0x21c
kernel: do_futex+0x168/0xb70
kernel: compat_SyS_futex+0x8c/0x174
kernel: __sys_trace_return+0x0/0x4
native: #00 pc 00018dac /system/lib/libc.so (_ZNK3jni13TypeSignatureIFvNS_6ObjectIN4mbgl7android6LatLngEEEdddEEclEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/type_signature.hpp:62)
native: #01 pc 000486c5 /system/lib/libc.so (operator()<jni::jobject *> /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:66)
native: #02 pc 006cec2d /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNSt6__ndk119__libcpp_mutex_lockEP15pthread_mutex_t /Volumes/Android/buildbot/src/android/ndk-r15-release/external/libcxx/include/__threading_support:65)
native: #03 pc 001d5487 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (lock_guard /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/__mutex_base:91)
native: #04 pc 0035a99f /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~FileSourceRequest /Users/ivo/git/mapbox-gl-native/src/mbgl/storage/file_source_request.cpp:18)
native: #05 pc 0035a9e5 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~FileSourceRequest /Users/ivo/git/mapbox-gl-native/src/mbgl/storage/file_source_request.cpp:13)
native: #06 pc 003cf7ad /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl12AsyncRequestEEclEPS2_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #07 pc 003cf72f /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~pair /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/utility:280)
native: #08 pc 003cf6ed /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNSt6__ndk116allocator_traitsINS_9allocatorINS_11__tree_nodeINS_12__value_typeINS_4pairIttEEN4mbgl12GlyphManager12GlyphRequestEEEPvEEEEE9__destroyINS4_IKS5_S8_EEEEvNS_17integral_constantIbLb0EEERSC_PT_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:1674)
native: #09 pc 003cf66b /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~__tree /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/__tree:1716)
native: #10 pc 003cf537 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~map /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/map:807)
native: #11 pc 003cf503 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~Entry /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../src/mbgl/text/glyph_manager.hpp:54)
native: #12 pc 003cf4d7 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~pair /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/utility:280)
native: #13 pc 003cf48f /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNSt6__ndk116allocator_traitsINS_9allocatorINS_11__hash_nodeINS_17__hash_value_typeINS_6vectorINS_12basic_stringIcNS_11char_traitsIcEENS1_IcEEEENS1_IS9_EEEEN4mbgl12GlyphManager5EntryEEEPvEEEEE9__destroyINS_4pairIKSB_SE_EEEEvNS_17integral_constantIbLb0EEERSI_PT_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:1674)
native: #14 pc 003cf39d /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~__hash_table /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/__hash_table:1480)
native: #15 pc 003cbbe1 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~unordered_map /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/__hash_table:1354)
native: #16 pc 003cbbbf /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~GlyphManager /Users/ivo/git/mapbox-gl-native/src/mbgl/text/glyph_manager.cpp:17)
native: #17 pc 00319f57 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl12GlyphManagerEEclEPS2_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #18 pc 00319fed /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~Impl /Users/ivo/git/mapbox-gl-native/src/mbgl/renderer/renderer_impl.cpp:62)
native: #19 pc 00316799 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl8Renderer4ImplEEclEPS3_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #20 pc 002d3191 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl8RendererEEclEPS2_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #21 pc 002d31c1 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~AndroidGLThread /Users/ivo/git/mapbox-gl-native/platform/android/src/android_gl_thread.hpp:29)
native: #22 pc 002ceb29 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl7android15AndroidGLThreadEEclEPS3_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #23 pc 001d8153 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~Impl /Users/ivo/git/mapbox-gl-native/src/mbgl/map/map.cpp:141)
native: #24 pc 001d8369 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~Impl /Users/ivo/git/mapbox-gl-native/src/mbgl/map/map.cpp:138)
native: #25 pc 001d83ef /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl3Map4ImplEEclEPS3_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #26 pc 000c3531 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl3MapEEclEPS2_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #27 pc 000c36c5 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (~NativeMapView /Users/ivo/git/mapbox-gl-native/platform/android/src/native_map_view.cpp:95)
native: #28 pc 00041605 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNKSt6__ndk114default_deleteIN4mbgl7android9LineLayerEEclEPS3_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:2540)
native: #29 pc 000d80b9 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (operator() /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:112)
native: #30 pc 000d8081 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (__invoke /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:110)
native: #31 pc 000d8125 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (operator()<jni::jobject *> /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:59)
native: #32 pc 000d80e5 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (__invoke<jni::jobject *> /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:55)
native: #33 pc 0000084b /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/oat/arm/base.odex (_ZNSt6__ndk17forwardIPN4mbgl8Response5ErrorEEEOT_RNS_16remove_referenceIS5_E4typeE /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/type_traits:2182)
at com.mapbox.mapboxsdk.maps.NativeMapView.nativeDestroy(Native method)
```
</details>
<details>
 <summary>DefaultFileSource thread blocked on `Scheduler::schedule()` (in `Mailbox::push`)</summary>

```
"DefaultFileSour" prio=1 tid=23 Native
| group="main" sCount=1 dsCount=0 flags=1 obj=0x139c3290 self=0xc8b60e00
| sysTid=12431 nice=19 cgrp=default sched=0/0 handle=0xc7f78970
| state=S schedstat=( 308846851 1260243511 1002 ) utm=21 stm=9 core=2 HZ=100
| stack=0xc7e7e000-0xc7e80000 stackSize=1006KB
| held mutexes=
kernel: __switch_to+0xb0/0xbc
kernel: futex_wait_queue_me+0xdc/0x168
kernel: futex_wait+0xf4/0x21c
kernel: do_futex+0x168/0xb70
kernel: compat_SyS_futex+0x8c/0x174
kernel: __sys_trace_return+0x0/0x4
native: #00 pc 00018dac /system/lib/libc.so (_ZNK3jni13TypeSignatureIFvNS_6ObjectIN4mbgl7android6LatLngEEEdddEEclEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/type_signature.hpp:62)
native: #01 pc 00048821 /system/lib/libc.so (_ZN3jni16MakeNativeMethodIZNS_17NativeMethodMakerIMZNS_30NativePeerMemberFunctionMethodIMN4mbgl7android9LineLayerEFvR7_JNIEnvxxEXadL_ZNS5_21setLineBlurTransitionES7_xxEEEclIS5_S5_vEEDaRKNS_5FieldIT0_xEEEUlS7_NS_6ObjectIS5_EExxE_KFvS7_SJ_xxEJEEclISK_EESC_PKcRKT_EUlPS6_PNS_7jobjectExxE_EEDaSQ_SQ_ST_PNSt6__ndk19enable_ifIXsr3std8is_classISR_EE5valueEvE4typeE /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:53)
native: #02 pc 006cec2d /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNSt6__ndk119__libcpp_mutex_lockEP15pthread_mutex_t /Volumes/Android/buildbot/src/android/ndk-r15-release/external/libcxx/include/__threading_support:65)
native: #03 pc 002d321b /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (lock_guard /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/__mutex_base:91)
native: #04 pc 001d5613 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl7Mailbox4pushENSt6__ndk110unique_ptrINS_7MessageENS1_14default_deleteIS3_EEEE /Users/ivo/git/mapbox-gl-native/src/mbgl/actor/mailbox.cpp:36)
native: #05 pc 0029b18b /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl8ActorRefINS_17FileSourceRequestEE6invokeIMS1_FvRKNS_8ResponseEEJS6_EEEvT_DpOT0_ /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/actor/actor_ref.hpp:34)
native: #06 pc 00299db5 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (operator() /Users/ivo/git/mapbox-gl-native/platform/default/default_file_source.cpp:118)
native: #07 pc 00296d11 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl17DefaultFileSource4Impl7requestEPNS_12AsyncRequestENS_8ResourceENS_8ActorRefINS_17FileSourceRequestEEE /Users/ivo/git/mapbox-gl-native/platform/default/default_file_source.cpp:153)
native: #08 pc 002a2e1d /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl11MessageImplINS_17DefaultFileSource4ImplEMS2_FvPNS_12AsyncRequestENS_8ResourceENS_8ActorRefINS_17FileSourceRequestEEEENSt6__ndk15tupleIJPS7_S5_S8_EEEE6invokeIJLj0ELj1ELj2EEEEvNSB_16integer_sequenceIjJXspT_EEEE /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/actor/message.hpp:34)
native: #09 pc 002a2d71 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl11MessageImplINS_17DefaultFileSource4ImplEMS2_FvPNS_12AsyncRequestENS_8ResourceENS_8ActorRefINS_17FileSourceRequestEEEENSt6__ndk15tupleIJPS7_S5_S8_EEEEclEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/actor/message.hpp:29)
native: #10 pc 001d58d5 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl7Mailbox7receiveEv /Users/ivo/git/mapbox-gl-native/src/mbgl/actor/mailbox.cpp:58)
native: #11 pc 001d5aaf /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl7Mailbox12maybeReceiveENSt6__ndk18weak_ptrIS0_EE /Users/ivo/git/mapbox-gl-native/src/mbgl/actor/mailbox.cpp:67)
native: #12 pc 002872ab /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl4util6ThreadINS0_5AlarmEE7receiveEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/util/thread.hpp:146)
native: #13 pc 002871cd /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (operator() /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/util/thread.hpp:136)
native: #14 pc 002a10b5 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl12WorkTaskImplIZNS_4util6ThreadINS_17DefaultFileSource4ImplEE8scheduleENSt6__ndk18weak_ptrINS_7MailboxEEEEUlvE_NS6_5tupleIJEEEE6invokeIJEEEvNS6_16integer_sequenceIjJXspT_EEEE /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/util/work_task_impl.hpp:43)
native: #15 pc 002a1057 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl12WorkTaskImplIZNS_4util6ThreadINS_17DefaultFileSource4ImplEE8scheduleENSt6__ndk18weak_ptrINS_7MailboxEEEEUlvE_NS6_5tupleIJEEEEclEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/util/work_task_impl.hpp:23)
native: #16 pc 002804f7 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl4util7RunLoop7processEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/util/run_loop.hpp:91)
native: #17 pc 00280339 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZN4mbgl4util7RunLoop3runEv /Users/ivo/git/mapbox-gl-native/platform/android/src/run_loop.cpp:233)
native: #18 pc 0029e865 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (operator() /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../include/mbgl/util/thread.hpp:56)
native: #19 pc 0029e587 /data/app/com.mapbox.mapboxsdk.testapp-fuKSTZvX1kTwbN8JYAGjoQ==/lib/arm/libmapbox-gl.so (_ZNSt6__ndk18__invokeIZN4mbgl4util6ThreadINS1_17DefaultFileSource4ImplEEC1IJRKNS_10shared_ptrINS1_10FileSourceEEERKNS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEERyEEESK_DpOT_EUlvE_JEEEDTclclsr3std6__ndk1E7forwardIT_Efp_Espclsr3std6__ndk1E7forwardIT0_Efp0_EEEOSQ_DpOSR_ /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/type_traits:4287)
native: #20 pc 00047f2f /system/lib/libc.so (_ZN3jni16MakeNativeMethodIZNS_30NativePeerMemberFunctionMethodIMN4mbgl7android9LineLayerEFNS_6ObjectINS_9ObjectTagEEER7_JNIEnvEXadL_ZNS4_13getLineOffsetES9_EEEclIS4_S4_vEEDaRKNS_5FieldIT0_xEEEUlS9_NS5_IS4_EEE_EEDaPKcRKT_ /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:122)
native: #21 pc 0001b0ef /system/lib/libc.so (_ZNK3jni13TypeSignatureINS_6ObjectIN4mbgl7android17TransitionOptionsEEEEclEv /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/type_signature.hpp:29)
(no managed stack frames)
```
</details>
<details>
 <summary>GL Thread blocked on it's mutex</summary>

```
"GLThread 1066" prio=5 tid=25 Waiting
| group="main" sCount=1 dsCount=0 flags=1 obj=0x139c3f28 self=0xc7393800
| sysTid=12467 nice=0 cgrp=default sched=0/0 handle=0xc787f970
| state=S schedstat=( 267480014 202200566 406 ) utm=22 stm=3 core=2 HZ=100
| stack=0xc777d000-0xc777f000 stackSize=1038KB
| held mutexes=
at java.lang.Object.wait(Native method)
- waiting on <0x08938807> (a android.opengl.GLSurfaceView$GLThreadManager)
at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1495)
- locked <0x08938807> (a android.opengl.GLSurfaceView$GLThreadManager)
at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1270)
```
</details>

<br/>
Not why it would be blocking there.

ivovandongen commented:
In https://github.com/mapbox/mapbox-gl-native/tree/ivd-android-glsurfaceview-3tier I've applied the mutex on the `UpdateParameters` to the 3-thread based solution btw. That is working very well and solved the last issues I was having there.

jfirebaugh commented:
The implementation should make a clearer division between responsibilities, mirroring the division between threads. None of the `NativeMapView` or `AndroidRendererFrontend` methods should be called from the GL thread.

* The implementation of `GLSurfaceView.Renderer` should be an independent class, `MapRenderer`, not `NativeMapView`. It will have a partially native implementation, like `NativeMapView`.
* `AndroidRendererFrontend::backend` → `MapRenderer::backend`
* `NativeMapView::nativeResizeFramebuffer`/`AndroidRendererFrontend::resizeFramebuffer` → `MapRenderer::nativeResizeFramebuffer`
* `NativeMapView::nativeRender`/`AndroidRendererFrontend::render` → `MapRenderer::nativeRender`

jfirebaugh commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/GlSurfaceViewRenderThread.java:
```diff
@@ -0,0 +1,25 @@
+ * {@link RenderThread} implementation that schedules using the
+ * {@link GLSurfaceView} thread.
+ */
+public class GlSurfaceViewRenderThread implements RenderThread {
```
Why so much indirection here? Can `NativeMapView` just have a `GLSurfaceView` reference?

ivovandongen replied:
This is in preparation of bringing back the TextureView (with a separate rendering thread).

jfirebaugh commented on platform/android/src/native_map_view.cpp:
```diff
@@ -70,117 +71,50 @@ NativeMapView::NativeMapView(jni::JNIEnv& _env,
+                                                                 *threadPool,
+                                                                 jni::Make<std::string>(_env,
+                                                                                        _programCacheDir),
+                                                                 [this] { this->requestRender(); },
```
Remove this indirection: arrange to pass a `jni::Object<GLSurfaceView>` directly, and have `AndroidRendererFrontend` call `requestRender` and `queueEvent` on it.

jfirebaugh commented on platform/android/src/android_gl_thread.hpp:
```diff
@@ -0,0 +1,74 @@
+namespace mbgl {
+namespace android {
+
+class AndroidGLThread : public Scheduler {
```
`AndroidGLThread` should also have direct access to the `GLSurfaceView` object. Implement `schedule` by invoking `queueEvent` via JNI, passing a new instance of a Java class which:

* Implements `Runnable`
* Has a native peer
* Native peer has a `std::weak_ptr<Mailbox> mailbox` member
* Implements `run` using native method that calls `Mailbox::maybeReceive(mailbox)`

Then you can remove `AndroidGLThread::process`, `AndroidGLThread::mutex`, `AndroidGLThread::queue`, `InvalidateCallback`, `NativeMapView::requestProcessing`, etc.

jfirebaugh commented:
The deadlock is caused by an invalid memory accessed due to order of operations in `AndroidGLThread`. You can see that the main thread is inside `~AndroidGLThread` -- in particular, it's in `~Renderer`. Since `std::unique_ptr<Renderer>` is the first member variable in `AndroidGLThread`, that means that the mutex has already been destroyed. The `DefaultFileSource` thread is trying to call `schedule` on a partially-destructed `AndroidGLThread` -- in particular, it's trying to obtain the lock for a destroyed mutex.

Changing the order of declarations such that `Renderer` is destroyed first may fix the issue, but I think the deeper refactoring suggested above will lead to a better design that doesn't have this issue. Probably `AndroidGLThread` shouldn't own the `Renderer` itself, but rather some other object should own both the `Renderer` and the `AndroidGLThread`.

ivovandongen commented:
@jfirebaugh Thanks for the review again. I've taken an initial stab at implementing your suggestions in https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/07fd0f51a3fa84f52746071bfb56ad9ffd92c8ca

Needs some work still and doesn't contain the work I did on re-initialization after backgrounding the app, but it works and renders.

jfirebaugh commented on platform/android/src/android_renderer_frontend.cpp:
```diff
@@ -1,72 +1,132 @@
+
+AndroidRendererFrontend::AndroidRendererFrontend(MapRenderer& mapRenderer_)
+        : mapRenderer(mapRenderer_)
+        , asyncInvalidate([&]() {
```
Do we need the async? I would expect `GLSurfaceView.requestRender` to coalesce internally.

ivovandongen replied:
No, you're right. Removed this. I left it in initially to skip some calls over the JNI boundary.

ivovandongen commented:
> Looks great! Did it fix the stability issues you were seeing too?

Still testing, but I don't run into the same deadlocks anymore. Looks good all-in-all.

I'm a little worried about the overhead involved in scheduling everything so granularly now as there are a lot of JNI calls (`queueEvent` and `requestRender`) are called a lot. This seems the most apparent when adding a large amount of annotations. I have not found a good way to measure this yet though, will spend some time on doing that.

ivovandongen commented on platform/android/src/android_renderer_frontend.cpp:
```diff
@@ -70,13 +71,13 @@ AndroidRendererFrontend::~AndroidRendererFrontend() = default;
-    mapRenderer.actor().invoke(&Renderer::setObserver, rendererObserver.get());
+    // Don't call the Renderer directly, but use MapRenderer#setObserver to make sure
+    // the Renderer may be re-initialised without losing the RendererObserver reference.
+    mapRenderer.setObserver(std::make_unique<ForwardingRendererObserver>(*mapRunLoop, observer));
```
@jfirebaugh This is a bit awkward, but I need the MapRenderer to be able to hold on to the Observer as the initialisation can happen in two orders (map observer first / gl surface initialisation first) and we need to deal with that using a mutex. Also, the MapRenderer needs the reference for re-initialisation after the app returns from the background so it can set it on the newly initialised renderer.

I kept the definition of the `ForwardingRendererObserver` as the `MapRenderer` has no access to the `RendererObserver` header.

ivovandongen commented:
@jfirebaugh I've fixed the problems with re-initialisation by ensuring that the context and the custom layers do not perform cleanup when the underlying gl context is already gone. On SurfaceView initialisation the MapRenderer flags the underlying context as lost in the `Context` through the `RendererBackend`. In the destructor of the `Context`, the cleanup of gl resources is skipped. Similarly, the `Renderer` is informed and in turn flags the context as lost on all `RenderCustomLayer`s. This flag is passed on to the custom layer's de-initialise function.

Perhaps this is not the cleanest api though.

A problem with resetting the Renderer like this is that the destruction of the `RenderLayer` used to trigger cleanup in the client's code (eg, destroying the client's object). This is now no longer so simple as the RenderLayer might be re-constructed. Maybe the best way to support this is to require an creation function from the client code instead of a pointer so that we can re-create when needed. Another option is to not call the deinitialise function when the context is lost, however, there might be other resources to cleanup on the client's code side.

I've worked around it in the Android example here: https://github.com/mapbox/mapbox-gl-native/blob/c534250ed9bceb6a4c28337a838851c4135eda73/platform/android/src/example_custom_layer.cpp#L104

zugaldia commented:
> Another complication is continued support of TextureView. Needed for certain use cases where mapviews are layered on top of other elements. Although we need to do the threading here ourselves, it would be great not to diverge too much for maintainability.

From [the docs](https://developer.android.com/reference/android/view/TextureView.html), the main reason to have a `TextureView` compatibility mode is that:

> Unlike SurfaceView, TextureView does not create a separate window but behaves as a regular View. This key difference allows a TextureView to be moved, transformed, animated, etc. For instance, you can make a TextureView semi-translucent by calling myView.setAlpha(0.5f).

A common request we've heard in the past is to have some degree of control over layering (e.g. map on top of video, or video on top of map, more in general, map on top of `View`, `View` on top of map). If we're able to provide examples of how to achieve this with a `GlSurfaceView` (e.g using `setZOrderMediaOverlay`/`setZOrderOnTop` or other mechanisms) this would accelerate deprecating `TextureView` mode. @tobrun @Guardiola31337 please let me know if I'm missing other important use cases.

Also, note that `textureMode()` is part of the public API - if we decided to remove it, that'd require a major SEMVER change (i.e. we'd need to wait until 6.x vs 5.2 to ship async rendering).

ivovandongen commented:
As discussed, changes to the custom layer api are backwards compatible. What has changed is:

- The initialise function might be called more than once - In case the app goes to the background and returns to the foreground, the de-initialise function is not called as the underlying GL context is gone and cleanup of any GL resources is no longer needed, instead the initialise function is called again so the layer can be prepared again.
- A new, optional, callback has been added that will be called before the initialise function if the underlying GL context was lost. 

This is only applicable for Android.

ivovandongen commented:
@zugaldia 

> Also, note that textureMode() is part of the public API - if we decided to remove it, that'd require a major SEMVER change (i.e. we'd need to wait until 6.x vs 5.2 to ship async rendering).

I suggest that we review and merge this PR then and if it's needed to ship in 5.2 / we really need to support TextureView for some use cases we add that in a follow up.

ivovandongen commented:
Noticed while testing some more that using the `GLSurfaceView` as the scheduler has one drawback. We need to destruct the `Renderer` on the GL thread, so on `MapView#onDestroy` we schedule a blocking call on the GL thread (through `~Map`->`RendererFrontend::reset`->`MapRenderer::reset`). When the onDestroy call is done after the GL thread has been paused for one reason or another, this blocks the main thread indefinitely -> ANR.

In most cases it's a matter of using the right callbacks: https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/3eb5be5834385a805ec86c3add29a999b0b487ed. In some others, it's less obvious and looks quite hackish/brittle: https://github.com/mapbox/mapbox-gl-native/pull/9576/commits/978525dca656616c8f8e8e92c3b44b1c9f410c12

zugaldia commented:
> In most cases it's a matter of using the right callbacks: 3eb5be5. In some others, it's less obvious and looks quite hackish/brittle: 978525d

Doesn't seem a blocker as the functionality is there. Instead, looks like a) in less obvious cases we could spend some follow-up work potentially improving the public APIs and, in any case, b) we need to make sure we document this very well beyond the Test App (public docs, examples). cc: @langsmith for the latter.

cc: @lilykaiser for visibility of the PR.

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java:
```diff
@@ -279,17 +281,36 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
+    glSurfaceView = (GLSurfaceView) findViewById(R.id.surfaceView);
+    glSurfaceView.setZOrderMediaOverlay(mapboxMapOptions.getRenderSurfaceOnTop());
+    glSurfaceView.setEGLContextClientVersion(2);
+    glSurfaceView.setEGLConfigChooser(new EGLConfigChooser());
```
Actually, this is how I started out (with a couple of variations). However, this didn't result in working context's on the two devices I have. This didn't give me confidence that it would work on the whole range of devices we support, that's why I ported the original to Java.
